### DISCUSSION TITLE
CPU usage time, simulation time and determinism

### DISCUSSION THREAD

**Discussion by bergerch** [shadow/shadow#1675] [Q&A] [answered]
As far as I understood Shadow can simulate the OS Kernel CPU usage and can capture processing time. For example consider an application process that creates 100 signatures (which consumes some time) then sends a messages to other processes.

This amount of time can be captured by Shadow's simulated OS Kernel, e.g. by to pushing back causally-depending events in the event queue by exactly this amount (correct me if I am wrong, but that was my impression on the influence of CPU usage on simulation time).

I was wondering how this time can be captured without disturbing the determinism of the simulation given that time measurements usually are a source of randomness?

**Comment by jtracey** [comment]
To the best of my knowledge, Shadow doesn't currently support CPU time simulation, so all computations in the simulation essentially take 0 sim time. If you need to run an experiment that requires taking CPU time into account, you have two options:

1. Use an emulator like mininet or netmirage instead, but be sure you have hardware that can handle running the full experiment in real time (so some kind of compute cluster with many CPUs).
2. First benchmark the computation time externally somehow (e.g., time how long it takes to generate 100 signatures many times, then take the median or mean), then model that extra computation by using something like sleep calls in the appropriate parts of the application code.

The first option has the advantage of modeling the actual, dynamic usage under deployment, the second option has the advantage of being deterministic (and running on cheaper hardware).

**Reply by bergerch** [reply]
I see.

The CPU time simulation was explained in an older talk about Shadow:
[https://youtu.be/oIrkFVTRNKs?t=2371](https://youtu.be/oIrkFVTRNKs?t=2371)
(note timestamp)
But I assume this has then changed by now?

I guess your second solution is a possible (but a little inconvenient) workaround to preserve the determinism while enabling to simulate cpu usage.

**Comment by robgjansen** [comment, marked as answer]
The Shadow `v1.x` design has a preliminary cpu modeling feature that attempted to take into account CPU usage of the plugins running in Shadow. It works roughly as follows:

- Shadow starts a timer before it jumps over to the plugin stack to continue running it.
- When the plugin executes a blocking syscall that requires us to jump back to the Shadow stack and continue progressing simulation time, we stop the timer.
- The _delay time_ is the difference between the timer start and stop time.
- Shadow adds the delay time to a CPU model for the host.
- If the cumulative delay is greater than a threshold, then the CPU is considered _blocked_ and will become _unblocked_ again after the simulation time advances by at least the delay time.
- Any event that would occur while the CPU is _blocked_ gets pushed back and is handled when the CPU becomes _unblocked_ again.

This CPU model was disabled by default, and could be enabled by setting `--cpu-threshold=N` to a positive value. As @bergerch points out, this causes simulations to become non-deterministic. We added a second feature to round the delay times to the nearest factor set by `--cpu-precision` to try to mitigate non-determinism, but this was a heuristic and does not guarantee determinism.

The problems with this CPU modeling approach is:
- It causes non-determinism as @bergerch points out, because different host CPUs will take different amounts of time to run the same workload.
- Does not model hosts with multiple CPUs operating in parallel (i.e., threads).
- The algorithm we used when a CPU entered the _blocked_ state, i.e., pushing all events backward in the queue until the CPU returned to the _unblocked_ state, destroyed the performance of the simulation. Basically, most of the simulation consisted in us moving events backward in the event queue rather than performing useful work.

So we disable the algorithm by default in Shadow `v1.x`, though the `v1.15.0` release does still allow you to turn on the algorithm as described above. (We do not have much experience with using the feature, so beware of that.)

Also, we have temporarily removed the feature altogether in Shadow `v2.x`. We hope one day to bring back CPU modeling with a much better design, but we've been busy with moving to the 2.0 architecture and have not made it that far yet. In the meantime, if you are using Shadow `v2.x` then @jtracey recommendation of making blocking sleep calls in your application would be a reasonable workaround.

**Reply by bergerch** [reply]
Hi,  thank you @robgjansen  very much for the detailed answer :)

First, I think the experimental feature in the earlier version sounds interesting. Maybe it is worth a trade-off to sacrifice determinism and thus exact reproducibility for the capability of enabling cpu usage simulation as long as re-played experiments produce results within an acceptable margin of error (e.g. consider a performance benchmark). Second, of course, determinism is very favorable. Consider a liveness bug in some blockchain protocol implementation which only appears under a certain pattern of messages exchanged. Then, we obviously would want to be able to reproduce that liveness violation.

Now, I am not sure if the rounding heuristic is promising because we would loose accuracy and, since measurements can still round in two ways, not be able to give strong guarantees. Would it be possible to "capture and reuse" measurements in subsequent simulation runs? Probably not easy as shadow doesnt really know what is going on in the application, so identifying repeating code executions and matching them with recorded measurements seems quite challenging.

As of know I also like the suggestion of @jtracey to use a workaround that can do the trick (should be feasable to model a certain small set of operations to capture most of the ongoing cpu usage anyways, at least in my case).

**Reply by robgjansen** [reply]
I agree the rounding heuristic is not great.

> Would it be possible to "capture and reuse" measurements in subsequent simulation runs?

As you mentioned, Shadow's general view of execution probably means it's not the right place to build this in. For specific use-cases, it might make sense to make the measurement from inside Shadow, but those changes should probably live in a fork. Doing something inside the application itself seems like the right place to me.

**Reply by bergerch** [reply]
Thanks for the explanation!

**Reply by lifeoftomi** [reply]
@robgjansen, trying to wrap my head around this and the original question. Is it correct to say that one could run an intensive application on Shadow and the host machine wouldn't feel a thing in terms of CPU usage? Also, how did the `shadow-plugin-bitcoin` experiment scale?

**Reply by bergerch** [reply]
@lifeoftomi 

- In terms of real-time (time that executing the simulation takes): it would be be quite time consuming and lead to a long simulation run as your CPU will be busy for quite some time.
- In terms of simulated-time (time that passes within the simulation): it wont be affected.

**Reply by lifeoftomi** [reply]
Thanks @bergerch, so could I run an application that would normally be CPU-intensive within Shadow and not experience heavy resource consumption on my host machine?

**Reply by bergerch** [reply]
@lifeoftomi No since Shadow runs the real application binaries. You would need to change the source code of your application (e.g. commenting out costly operations) and re-compile. But this might easily break the application (consider the generation of signatures: if you leave that away then meassages can not authenticate succesfully and will not be processed at the receiving side). Proof-of-Work might be an exception here, because it is not doing something "useful". In a simulation, PoW mechanics could be easily replaced by just wating for a random amount of time instead of hashing and verifying the found nonces. But you would need to change the source code of the PoW protocol in the first place..

edit: I did not read the paper about the Bitcoin simulation, but I can not imagine that they would actually execute the hashing of PoW on a host because this can be easily modeled by some stochastic process and thus abstracted away from the application, saving CPU time

**Reply by robgjansen** [reply]
Let's redirect discussion about bitcoin scalability to #1753.

**Comment by alrevuelta** [comment]
> This CPU model was disabled by default, and could be enabled by setting --cpu-threshold=N to a positive value. As @bergerch points out, this causes simulations to become non-deterministic. We added a second feature to round the delay times to the nearest factor set by --cpu-precision to try to mitigate non-determinism, but this was a heuristic and does not guarantee determinism.

Can see `cpu-threshold` flag was removed in https://github.com/shadow/shadow/pull/1263/files, @robgjansen is there a way to configure this in `> v3.0.0`? Or more in general, any new updates on how to run in the latest versions taking CPU consumption time into account?

**Reply by robgjansen** [reply]
With all of the work involved with migrating to Rust, we have not been able to implement a proper CPU modeling system yet. It is still on our [list of things we want to get funded](https://github.com/shadow/shadow/discussions/2995).

Some people still have some particular function(s) inside of an application that you expect to take a lot of CPU time and you would like to account for it when running inside of Shadow. In this case, as a workaround we have been recommending that some measurements of the CPU-intensive functions be taken outside of Shadow on bare-metal if possible, and then those measurements could be converted into periodic calls to `sleep()` or `nanosleep()` in your application. Those syscalls will cause shadow to block your application thread for the given amount of time.

Sorry, I realize it would be much nicer if this was just handled automatically on the Shadow side so that every application didn't have to go through this extra step. We just haven't quite gotten there yet.

In any case, hope that helps you understand the current situation.

**Comment by diegomrsantos** [comment]
Hi all — In the context of *Gossipsub* behaviour of an Ethereum consensus client inside Shadow, consider the following:

### Scenario

1. **Network setup**  
   * A → B one-way latency = **50 ms** (fixed, no jitter).  
   * At `t=150 ms` (simulated) B receives a beacon-block message **M** from A.

2. **CPU-bound validation on B**  
   * Immediately after reception, B runs `MessageValidator::validate()`  
   * Validation costs **≈10 ms wall-clock** but makes **no blocking syscalls**.  
   * Because Shadow “models the CPU as infinitely fast”, B’s virtual time stays **frozen at 150 ms** while this runs. 

3. **Concurrent control message**  
   * At real ~153 ms another peer sends B an **`IDONTWANT`(M)**.  
   * Shadow timestamps that packet **153 ms** but queues it—B’s clock is still 150 ms, and the thread hasn’t yielded.  

4. **Two possible exits from validation**  

   | Path A – forward immediately | Path B – no forward (just blocks) |
   |---|---|
   | `validate()` returns and the code calls `send()` to forward **M** right away | `validate()` returns and the code blocks in `poll()`/`recv()` |

### My understanding of Shadow’s rules
* **Rule 1** – A host’s virtual clock only advances when the plugin thread makes a *blocking* or *time-related* syscall. 
* **Rule 2** – On each syscall boundary Shadow **first dequeues every pending event whose `timestamp ≤ current_time`** and delivers them to the host; events with later timestamps stay queued.

### Questions

1. **Delivery order at the syscall boundary**  
   Given Rule 2, it sounds like the queued `IDONTWANT @ 153 ms` is **not** delivered when B calls `send()` in *Path A* because `153 ms > 150 ms`.  
   *⇒ The forward would be scheduled at 150 ms and race ahead of the control message.*  
   Is that indeed Shadow’s behaviour?

2. **Charging CPU time to avoid the race**  
   If we insert an explicit  
   ```rust
   std::thread::sleep(Duration::from_millis(10));
   ```
 
or enable model_unblocked_syscall_latency) inside the validator, B’s virtual clock should advance to 160 ms before the next syscall, causing the queued IDONTWANT @ 153 ms to be delivered first, suppressing the forward — i.e., matching production Gossipsub semantics.
⇒ Is this the recommended way to model validator latency, or is there a better approach?

Clarifying this will help us avoid over-counting redundant block forwards in large-scale Ethereum network experiments. Thanks a lot!

**Reply by sporksmith** [reply]
> At real ~153 ms another peer sends B an IDONTWANT(M).

I'm not sure what "real" means here. I'll assume below that you're talking about simulation time t=153ms.

> Rule 1 – A host’s virtual clock only advances when the plugin thread makes a blocking or time-related syscall.

I would describe it as only blocking, counting e.g. `sleep` as blocking. Other time related syscalls like `gettimeofday` do not advance the clock.

(And of course this rule is broken if `model_unblocked_syscall_latency` is enabled)

> Rule 2 – On each syscall boundary Shadow first dequeues every pending event whose timestamp ≤ current_time and delivers them to the host; events with later timestamps stay queued.

Not quite, but probably close enough for your purposes. To clarify though:

* There should never be an event scheduled whose timestamp is < than the current time. e.g. one way that we ensure this is that parallel worker threads are only allowed to execute until the end of the current "round" of simulated time, and the round size is chosen based on link latency such that a packet from another host can never be received "in the past".
* When potentially moving time forward, we first check if this would move time beyond the next scheduled event. If so, then we yield and schedule ourselves to run again at the new time. This only happens when modeling unblocked syscall latency is enabled. Otherwise we never move time forward "inline"; when a syscall is blocking (e.g. a blocking read, or a sleep, etc), we just stop executing. Some future event (e.g. data arriving on the file descriptor being read-from) generally continues execution of the thread.

> Given Rule 2, it sounds like the queued IDONTWANT @ 153 ms is not delivered when B calls send() in Path A because 153 ms > 150 ms.
>  ⇒ The forward would be scheduled at 150 ms and race ahead of the control message.

Yes, I believe this is correct. (If the message was *sent* at 153 ms, then the *arrival* would actually be at 153ms + link latency, but this doesn't change the answer)

>  If we insert an explicit `std::thread::sleep(Duration::from_millis(10));` or enable model_unblocked_syscall_latency) inside the validator, B’s virtual clock should advance to 160 ms before the next syscall, causing the queued IDONTWANT @ 153 ms to be delivered first, suppressing the forward — i.e., matching production Gossipsub semantics.
⇒ Is this the recommended way to model validator latency, or is there a better approach?

Either approach *could* work. I would lean towards adding the explicit sleep in this case since 10ms is a pretty significant delay, and I think `model_unblocked_syscall_latency` will end up giving you a much smaller one; (or possibly none at all unless you also override [`max_unapplied_cpu_latency`](https://shadow.github.io/docs/guide/shadow_config_spec.html#experimentalmax_unapplied_cpu_latency)  to ~0).

**Reply by sporksmith** [reply]
It's also worth mentioning that the `model_unblocked_syscall_latency` feature as currently implemented is meant primarily to allow escaping busy-loops, while trying to otherwise minimize the effect on actual simulation outcome. e.g. a change in its default behavior that I'm considering (https://github.com/shadow/shadow/issues/2161#issuecomment-2941905361) might cause surprises if you rely on it for this. (Though if we do go forward with such a change it'd at least be noted in the changelog, and could still be overridden in the config)
